pilot/pkg/bootstrap/server.go

```diff
@@ -813,22 +835,101 @@ func (s *Server) initAdmissionController(args *PilotArgs) error {
 	return nil
 }
 
-// NewDiscoveryServer creates DiscoveryServer that sources data from Pilot's internal mesh data structures
 func (s *Server) initGrpcServer() {
-	// TODO for now use hard coded / default gRPC options. The constructor may evolve to use interfaces that guide specific options later.
-	// Example:
-	//		grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(uint32(someconfig.MaxConcurrentStreams)))
-	var grpcOptions []grpc.ServerOption
+	grpcOptions := s.grpcServerOptions()
+	s.GRPCServer = grpc.NewServer(grpcOptions...)
+}
 
-	var interceptors []grpc.UnaryServerInterceptor
+// The secure grpc will start when the credentials are found.
+func (s *Server) secureGrpcStart(listener net.Listener) {
+	certDir := os.Getenv("PILOT_CERT_DIR")
+	if certDir == "" {
+		certDir = PilotCertDir // /etc/certs
+	}
+	if !strings.HasSuffix(certDir, "/") {
+		certDir = certDir + "/"
+	}
 
-	// TODO: log request interceptor if debug enabled.
+	for i := 0; i < 10; i++ {
+		opts := s.grpcServerOptions()
 
-	// setup server prometheus monitoring (as final interceptor in chain)
-	interceptors = append(interceptors, prometheus.UnaryServerInterceptor)
-	prometheus.EnableHandlingTimeHistogram()
+		// This is used for the grpc h2 implementation. It doesn't appear to be needed in
+		// the case of golang h2 stack.
+		creds, err := credentials.NewServerTLSFromFile(certDir+model.CertChainFilename,
+			certDir+model.KeyFilename)
+		// certs not ready yet.
+		if err != nil {
+			time.Sleep(5 * time.Second)
```

Comment:
This seems arbitrary. May be panic and fail or simply retry twice and fail

Reply:
In the flaky tests we've seen >1 min delay until the cert shows up. I can make it configurable ?

Reply:
Sure.. with a bound - thats all I am asking.. else, some installation might be perpetually spinning waiting for certs, without giving any proper message to user. A panic after 4 minutes is easily noticeable..

Reply:
Right now it's 50 seconds - but I don't think panic is right unless controlPlaneAuthPolicy is on.